File: backend/src/main/java/com/eps/procurement/policy/ProcurementPolicy.java
package com.eps.procurement.policy;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Procurement policy — pure rules, no state. Mirrors the frontend policy module
 * (src/store/procurementPolicy.js) so the same guarantees hold server side:
 *
 * <ol>
 *   <li>Item classification: a laptop can only ever be raised under
 *       Equipment &amp; Assets → Laptop (and likewise for other known items).</li>
 *   <li>Department routing: which team owns a category and may therefore source,
 *       award and process purchase orders for it.</li>
 *   <li>The purchase order state machine.</li>
 * </ol>
 */
public final class ProcurementPolicy {

    private ProcurementPolicy() {}

    // ── 1. Item classification ───────────────────────────────
    public record ItemRule(String key, List<String> keywords, String category, String subcategory) {}

    public static final List<ItemRule> ITEM_RULES = List.of(
        new ItemRule("laptop", List.of("laptop", "macbook", "notebook pc", "thinkpad", "ideapad", "ultrabook", "chromebook"), "Equipment & Assets", "Laptop"),
        new ItemRule("desktop", List.of("desktop", "workstation", "all-in-one pc", "tower pc", "imac"), "Equipment & Assets", "Desktop"),
        new ItemRule("monitor", List.of("monitor", "ultrasharp", "display panel"), "Equipment & Assets", "Monitor"),
        new ItemRule("keyboard", List.of("keyboard"), "Equipment & Assets", "Keyboard"),
        new ItemRule("mouse", List.of("mouse", "trackpad"), "Equipment & Assets", "Mouse"),
        new ItemRule("headset", List.of("headset", "headphone", "earphone"), "Equipment & Assets", "Headset"),
        new ItemRule("webcam", List.of("webcam"), "Equipment & Assets", "Webcam"),
        new ItemRule("dock", List.of("docking station", "dock hub"), "Equipment & Assets", "Docking Station"),
        new ItemRule("license", List.of("license", "licence"), "Software & Digital Services", "Software License"),
        new ItemRule("saas", List.of("subscription", "saas"), "Software & Digital Services", "SaaS Subscription"),
        new ItemRule("cloud", List.of("cloud credits", "cloud service", "aws ", "azure", "gcp"), "Software & Digital Services", "Cloud Service"),
        new ItemRule("furniture", List.of("chair", "desk", "table", "furniture", "cabinet", "workbench"), "Facilities", "Furniture"),
        new ItemRule("renovation", List.of("renovation", "remodel", "civil work", "interior work"), "Facilities", "Renovation"),
        new ItemRule("repairs", List.of("repair", "servicing", "maintenance visit"), "Facilities", "Repairs"),
        new ItemRule("cleaning", List.of("cleaning", "housekeeping", "sanitiser", "sanitizer"), "Facilities", "Cleaning Supplies")
    );

    /** The classification rule an item title falls under, or {@code null}. */
    public static ItemRule classifyItem(String title) {
        String haystack = " " + (title == null ? "" : title.toLowerCase()) + " ";
        for (ItemRule rule : ITEM_RULES) {
            for (String keyword : rule.keywords()) {
                if (haystack.contains(keyword)) {
                    return rule;
                }
            }
        }
        return null;
    }

    public static ItemRule suggestCategory(String title) {
        return classifyItem(title);
    }

    /**
     * Validates the category/subcategory chosen for an item title.
     *
     * @return {@code null} when the combination is legal, otherwise a human readable reason.
     */
    public static String validateItemCategory(String title, String category, String subcategory) {
        ItemRule rule = classifyItem(title);
        if (rule == null) {
            return null;
        }
        boolean sameCategory = rule.category().equalsIgnoreCase(category == null ? "" : category.trim());
        boolean sameSubcategory = rule.subcategory().equalsIgnoreCase(subcategory == null ? "" : subcategory.trim());
        if (sameCategory && sameSubcategory) {
            return null;
        }
        return "\"" + (title == null ? "" : title.trim()) + "\" is a " + rule.subcategory().toLowerCase()
                + " item. It must be raised under " + rule.category() + " -> " + rule.subcategory()
                + " (selected: " + (category == null || category.isBlank() ? "-" : category) + " -> "
                + (subcategory == null || subcategory.isBlank() ? "-" : subcategory) + ").";
    }

    // ── 2. Department routing ────────────────────────────────
    public static final Map<String, String> CATEGORY_TEAMS = Map.of(
        "Equipment & Assets", "equipment_team",
        "Software & Digital Services", "software_team",
        "Facilities", "facilities_team"
    );

    public static final Map<String, String> TEAM_LABELS;
    static {
        Map<String, String> labels = new LinkedHashMap<>();
        labels.put("equipment_team", "Equipment Team");
        labels.put("software_team", "IT / Software Team");
        labels.put("facilities_team", "Facilities Team");
        labels.put("procurement_officer", "Central Procurement");
        labels.put("admin", "Administrator");
        TEAM_LABELS = Map.copyOf(labels);
    }

    public static final List<String> DEPARTMENT_TEAM_ROLES =
            List.of("equipment_team", "software_team", "facilities_team");

    /** Roles that may process every category. */
    public static final List<String> CENTRAL_ROLES = List.of("procurement_officer", "admin");

    public static String teamForCategory(String category) {
        if (category == null) {
            return "procurement_officer";
        }
        return CATEGORY_TEAMS.getOrDefault(category.trim(), "procurement_officer");
    }

    public static String teamLabel(String role) {
        return TEAM_LABELS.getOrDefault(role, role);
    }

    /**
     * Central procurement handles everything; a department team only handles the
     * category it is designated for (IT → software, Facilities → facilities,
     * Equipment → equipment).
     */
    public static boolean canProcessCategory(String role, String category) {
        if (role == null) {
            return false;
        }
        return CENTRAL_ROLES.contains(role) || teamForCategory(category).equals(role);
    }

    public static String categoryDenialReason(String role, String category) {
        return teamLabel(role) + " is not designated for " + category + ". "
                + teamLabel(teamForCategory(category)) + " or central procurement must process it.";
    }

    // ── 3. Purchase order state machine ─────────────────────
    public static final Map<String, List<String>> PO_FLOW;
    static {
        Map<String, List<String>> flow = new LinkedHashMap<>();
        flow.put("draft", List.of("sent", "cancelled"));
        flow.put("sent", List.of("accepted", "cancelled"));
        flow.put("accepted", List.of("in_transit", "cancelled"));
        flow.put("in_transit", List.of("delivered"));
        flow.put("delivered", List.of("closed"));
        flow.put("closed", List.of());
        flow.put("cancelled", List.of());
        PO_FLOW = Map.copyOf(flow);
    }

    public static final List<String> PO_STAGE_ORDER =
            List.of("draft", "sent", "accepted", "in_transit", "delivered", "closed");

    public static List<String> nextPoStatuses(String status) {
        return PO_FLOW.getOrDefault(status == null ? "" : status.toLowerCase(), List.of());
    }

    public static boolean canTransitionPo(String from, String to) {
        return to != null && nextPoStatuses(from).contains(to.toLowerCase());
    }
}

File: front-end/src/store/procurementPolicy.js
/**
 * ============================================================
 * Procurement policy — pure rules, no state
 * ============================================================
 * Everything here is a pure function so the workflow rules can be
 * unit / end-to-end tested without React or the store.
 *
 *  1. Item classification rules — a laptop can only ever be raised
 *     under Equipment & Assets → Laptop, and the same guard applies
 *     to every other well-known item type.
 *  2. Department routing — which team owns a category and is therefore
 *     allowed to select the vendor and process the purchase order.
 *  3. Purchase order state machine.
 */

// ── 1. Item classification ───────────────────────────────────
/**
 * Keyword → mandatory (category, subcategory) pairs.
 * The first rule whose keyword appears in the item title wins.
 */
export const ITEM_CATEGORY_RULES = [
  { key: 'laptop', keywords: ['laptop', 'macbook', 'notebook pc', 'thinkpad', 'ideapad', 'ultrabook', 'chromebook'], category: 'Equipment & Assets', subcategory: 'Laptop' },
  { key: 'desktop', keywords: ['desktop', 'workstation', 'all-in-one pc', 'tower pc', 'imac'], category: 'Equipment & Assets', subcategory: 'Desktop' },
  { key: 'monitor', keywords: ['monitor', 'ultrasharp', 'display panel'], category: 'Equipment & Assets', subcategory: 'Monitor' },
  { key: 'keyboard', keywords: ['keyboard'], category: 'Equipment & Assets', subcategory: 'Keyboard' },
  { key: 'mouse', keywords: ['mouse', 'trackpad'], category: 'Equipment & Assets', subcategory: 'Mouse' },
  { key: 'headset', keywords: ['headset', 'headphone', 'earphone'], category: 'Equipment & Assets', subcategory: 'Headset' },
  { key: 'webcam', keywords: ['webcam'], category: 'Equipment & Assets', subcategory: 'Webcam' },
  { key: 'dock', keywords: ['docking station', 'dock hub'], category: 'Equipment & Assets', subcategory: 'Docking Station' },
  { key: 'license', keywords: ['license', 'licence'], category: 'Software & Digital Services', subcategory: 'Software License' },
  { key: 'saas', keywords: ['subscription', 'saas'], category: 'Software & Digital Services', subcategory: 'SaaS Subscription' },
  { key: 'cloud', keywords: ['cloud credits', 'cloud service', 'aws ', 'azure', 'gcp'], category: 'Software & Digital Services', subcategory: 'Cloud Service' },
  { key: 'furniture', keywords: ['chair', 'desk', 'table', 'furniture', 'cabinet', 'workbench'], category: 'Facilities', subcategory: 'Furniture' },
  { key: 'renovation', keywords: ['renovation', 'remodel', 'civil work', 'interior work'], category: 'Facilities', subcategory: 'Renovation' },
  { key: 'repairs', keywords: ['repair', 'servicing', 'maintenance visit'], category: 'Facilities', subcategory: 'Repairs' },
  { key: 'cleaning', keywords: ['cleaning', 'housekeeping', 'sanitiser', 'sanitizer'], category: 'Facilities', subcategory: 'Cleaning Supplies' },
];

const normalise = (value) => ` ${String(value || '').toLowerCase()} `;

/** The classification rule an item title falls under, or null. */
export const classifyItem = (title) => {
  const haystack = normalise(title);
  return ITEM_CATEGORY_RULES.find((rule) => rule.keywords.some((k) => haystack.includes(k))) || null;
};

/**
 * Validate the category/subcategory chosen for an item.
 * Returns null when valid, otherwise a human readable message.
 * A laptop request, for instance, can only be filed under
 * Equipment & Assets → Laptop.
 */
export const validateItemCategory = ({ title, category, subcategory }) => {
  const rule = classifyItem(title);
  if (!rule) return null;
  if (category !== rule.category || subcategory !== rule.subcategory) {
    return `"${String(title).trim()}" is a ${rule.subcategory.toLowerCase()} item. It must be raised under ${rule.category} → ${rule.subcategory} (selected: ${category || '—'} → ${subcategory || '—'}).`;
  }
  return null;
};

/** Suggested category/subcategory for a title, for inline hints in the UI. */
export const suggestCategory = (title) => {
  const rule = classifyItem(title);
  return rule ? { category: rule.category, subcategory: rule.subcategory } : null;
};

// ── 2. Department routing ────────────────────────────────────
/** Which department team owns each procurement category. */
export const CATEGORY_TEAMS = {
  'Equipment & Assets': 'equipment_team',
  'Software & Digital Services': 'software_team',
  Facilities: 'facilities_team',
};

export const TEAM_LABELS = {
  equipment_team: 'Equipment Team',
  software_team: 'IT / Software Team',
  facilities_team: 'Facilities Team',
  procurement_officer: 'Central Procurement',
  admin: 'Administrator',
};

export const TEAM_DEPARTMENTS = {
  equipment_team: 'Procurement',
  software_team: 'IT',
  facilities_team: 'Facilities',
};

export const DEPARTMENT_TEAM_ROLES = ['equipment_team', 'software_team', 'facilities_team'];
export const CENTRAL_ROLES = ['procurement_officer', 'admin'];

export const teamForCategory = (category) => CATEGORY_TEAMS[category] || 'procurement_officer';

/**
 * Can this role source / award / process work in a category?
 * Central procurement handles everything; a department team only
 * handles the category it is designated for (IT → software,
 * Facilities → facilities, Equipment → equipment).
 */
export const canProcessCategory = (role, category) =>
  CENTRAL_ROLES.includes(role) || teamForCategory(category) === role;

export const assertCanProcessCategory = (role, category) => {
  if (!canProcessCategory(role, category)) {
    throw new Error(
      `Your team is not designated for ${category}. ${TEAM_LABELS[teamForCategory(category)]} or central procurement must process it.`,
    );
  }
};

// ── 3. Purchase order state machine ─────────────────────────
/**
 * A purchase order is raised on vendor award and first goes to finance
 * for approval. Finance approval hands it back to the procurement
 * officer (or the designated department team), who issues it to the
 * supplier for further processing.
 */
export const PO_FLOW = {
  draft: ['pending_finance', 'cancelled'],
  pending_finance: ['finance_approved', 'finance_rejected', 'cancelled'],
  finance_rejected: ['pending_finance', 'cancelled'],
  finance_approved: ['sent', 'cancelled'],
  sent: ['accepted', 'cancelled'],
  accepted: ['in_transit', 'cancelled'],
  in_transit: ['delivered'],
  delivered: ['closed'],
  closed: [],
  cancelled: [],
};

export const PO_ACTION_LABELS = {
  pending_finance: 'Send to Finance for Approval',
  finance_approved: 'Approve Purchase Order (Finance)',
  finance_rejected: 'Reject Purchase Order (Finance)',
  sent: 'Issue to Supplier',
  accepted: 'Record Supplier Acknowledgement',
  in_transit: 'Mark Dispatched',
  delivered: 'Mark Delivered',
  closed: 'Close Order',
  cancelled: 'Cancel Order',
};

export const PO_STAGE_ORDER = [
  'draft',
  'pending_finance',
  'finance_approved',
  'sent',
  'accepted',
  'in_transit',
  'delivered',
  'closed',
];

/**
 * Which roles may drive a PO into a given status. Finance owns the
 * approval gate; procurement (central or the designated department team)
 * owns everything else, including issuing the approved order to the
 * supplier.
 */
export const PO_TRANSITION_ROLES = {
  pending_finance: ['procurement_officer', 'equipment_team', 'software_team', 'facilities_team', 'admin'],
  finance_approved: ['finance_officer', 'admin'],
  finance_rejected: ['finance_officer', 'admin'],
  sent: ['procurement_officer', 'equipment_team', 'software_team', 'facilities_team', 'admin'],
};

/** Finance-only stages: procurement can view but not move them. */
export const FINANCE_PO_STATUSES = ['pending_finance'];

export const rolesForPoTransition = (to) => PO_TRANSITION_ROLES[to] || null;

/** Does this role own the given PO transition? */
export const canRoleTransitionPo = (role, to) => {
  const allowed = rolesForPoTransition(to);
  return !allowed || allowed.includes(role);
};

export const nextPoStatuses = (status) => PO_FLOW[status] || [];

export const canTransitionPo = (from, to) => nextPoStatuses(from).includes(to);

